File: java/src/main/java/benchmarks/BWTHuffDecode.java
package benchmarks;

import java.util.*;

public class BWTHuffDecode extends BWTHuffEncode {

    private CompressedData compressedData;
    private byte[] decompressed;

    public BWTHuffDecode() {
        sizeVal = configVal("size");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "BWTHuffDecode";
    }

    @Override
    public void prepare() {
        testData = generateTestData(sizeVal);
        compressedData = compress(testData);
    }

    @Override
    public void run(int iterationId) {
        decompressed = decompress(compressedData);
        resultVal += decompressed.length;
    }

    @Override
    public long checksum() {
        long res = resultVal;
        if (Arrays.equals(testData, decompressed)) {
            res += 1000000;
        }
        return res;
    }
}

File: kotlin/src/main/kotlin/benchmarks/BWTHuffDecode.kt
package benchmarks

import java.util.*

class BWTHuffDecode : BWTHuffEncode() {
    private lateinit var compressedData: CompressedData
    private lateinit var decompressed: ByteArray

    init {
    }

    override fun name(): String = "BWTHuffDecode"

    override fun prepare() {
        testData = generateTestData(sizeVal)
        compressedData = compress(testData)
    }

    override fun run(iterationId: Int) {
        decompressed = decompress(compressedData)
        resultVal += decompressed.size.toUInt()
    }

    override fun checksum(): UInt {
        var res = resultVal
        if (testData.contentEquals(decompressed)) {
            res += 1000000u
        }
        return res
    }
}
